File: problem/bool-and-or/6/main.lua
-- main.lua: main file for solution to the bool-and-or problem
--
-- Copyright (C) 2020-2021  Rudy Matela
-- All rights reserved.
--
--
-- This file is part of the CScx Online Judge.
--
-- This is appended to the submitted Lua program and tests the negation function.
--
-- The submitted file processes the standard input and this processes the
-- "in.txt" file.

function words (str)
  if str
  then
    local r={}
    for s in string.gmatch(str, "([^%s]+)") do
      table.insert(r, s)
    end
    return table.unpack(r)
  else
    return nil
  end
end

io.input("in.txt")
p, o, q = words(io.read("*line"))
while p and o and q do
  p = p == "true"
  o = o == "and" and conjunction or disjunction
  q = q == "true"
  print(o(p,q))
  p, o, q = words(io.read("*line"))
end

File: problem/bool-and-or/6/main.cc
/*
 * main.c: main file for solutions to the "bool-and-or" exercise
 *
 * This is to be linked to the submitted file.
 * It processes values from standard input then from the "in.txt" file.
 *
 *
 * Copyright (C) 2020-2021  Rudy Matela
 * All rights reserved.
 */
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cstdlib>
bool conjunction(bool,bool);
bool disjunction(bool,bool);
void process(FILE *fi)
{
    char sp[6], so[4], sq[6];
    bool p, q;
    bool (*o)(bool,bool);
    while (fscanf(fi, " %s %s %s", sp, so, sq)==3) {
        p = strcmp(sp, "true") == 0;
        o = strcmp(so, "and") == 0 ? conjunction : disjunction;
        q = strcmp(sq, "true") == 0;
        puts(o(p,q) ? "true" : "false");
    }
}
void errxit(const char *msg)
{
	fprintf(stderr, msg);
	exit(1);
}
int main()
{
	FILE *f;
	process(stdin);
	f = fopen("in.txt", "r");
	if (!f) errxit("could not open in.txt");
	process(f);
	fclose(f);
	return 0;
}
